### PR TITLE
Enable pipeline mode for write and restore from latest backup

### PR DESCRIPTION
**Pull request by rohitjoshi**
Exposing FFI function `enable_pipelined_write` to Rust interface.

### CODE REVIEW

**Comment by iSynaptic** [comment]
@rohitjoshi CI builds now have `rustfmt` checks. After you make your changes, you may need to run `cargo fmt --all` before this can be merged to master.

File: src/db_options.rs [hunk 1]
```diff
@@ -177,6 +176,20 @@ impl Options {
...
+    /// maintained for WAL write and memtable write. A write thread first enter WAL
+    /// writer queue and then memtable writer queue. Pending thread on the WAL
+    /// writer queue thus only have to wait for previous writers to finish their
+    /// WAL writing but not the memtable writing. Enabling the feature may improve
+    /// write throughput and reduce latency of the prepare phase of two-phase
+    /// commit. db_bench benchmark shows 20% write throughput improvement
+    /// with concurrent writers and WAL enabled
+    pub fn enable_pipelined_write(&self, v: bool) {
+        unsafe {
+            ffi::rocksdb_options_enable_pipelined_write(self.inner, v as c_uchar);
```
Comment (aleksuss) [inline src/db_options.rs]: A correct name of the func is `rocksdb_options_set_enable_pipelined_write`.

**Comment by aleksuss** [comment]
@rohitjoshi @iSynaptic FYI: A call of the function `rocksdb_options_set_enable_pipelined_write` requires a version rocksdb >= 5.16, therefore we should update a submodule rocksdb as well.

P.S. Current version of pre-compiled rocksdb lib on my mac is 5.14.3. So if we'll merge this PR we have to compile rocksdb dependency from sources.

File: src/db_options.rs [hunk 2]
```diff
@@ -177,6 +176,20 @@ impl Options {
...
 
+    /// If enable_pipelined_write is true, separate write thread queue is
+    /// maintained for WAL write and memtable write. A write thread first enter WAL
+    /// writer queue and then memtable writer queue. Pending thread on the WAL
+    /// writer queue thus only have to wait for previous writers to finish their
+    /// WAL writing but not the memtable writing. Enabling the feature may improve
+    /// write throughput and reduce latency of the prepare phase of two-phase
+    /// commit. db_bench benchmark shows 20% write throughput improvement
+    /// with concurrent writers and WAL enabled
+    pub fn enable_pipelined_write(&self, v: bool) {
```
Comment (aleksuss) [inline src/db_options.rs]:
```suggestion
    pub fn enable_pipelined_write(&mut self, v: bool) {
```

File: src/db_options.rs [hunk 3]
```diff
@@ -177,6 +176,20 @@ impl Options {
...
     }
 
+    /// If enable_pipelined_write is true, separate write thread queue is
+    /// maintained for WAL write and memtable write. A write thread first enter WAL
+    /// writer queue and then memtable writer queue. Pending thread on the WAL
+    /// writer queue thus only have to wait for previous writers to finish their
+    /// WAL writing but not the memtable writing. Enabling the feature may improve
+    /// write throughput and reduce latency of the prepare phase of two-phase
+    /// commit. db_bench benchmark shows 20% write throughput improvement
+    /// with concurrent writers and WAL enabled
```
Comment (aleksuss) [inline src/db_options.rs]: Add example please like this
```rust
    /// Default: `false`
    ///
    /// # Example
    ///
    /// ```
    /// use rocksdb::Options;
    ///
    /// let mut opts = Options::default();
    /// opts.enable_pipelined_write(true);
    /// ```
```

**Comment by iSynaptic** [comment]
> 
> 
> @rohitjoshi @iSynaptic FYI: A call of the function `rocksdb_options_set_enable_pipelined_write` requires a version rocksdb >= 5.16, therefore we should update a submodule rocksdb as well.
> 
> P.S. Current version of pre-compiled rocksdb lib on my mac is 5.14.3. So if we'll merge this PR we have to compile rocksdb dependency from sources.

@aleksuss I reassigned #217 to you since you are planning on doing the upgrade.

**Comment by iSynaptic** [comment]
@rohitjoshi @aleksuss `master` now has 5.17.2.

**Comment by rohitjoshi** [comment]
@iSynaptic I will close this PR and submit two separate
1. Enable pipeline  write
2. Restore from the back.

**Comment by iSynaptic** [comment]
Closing this PR since it's being split into two different PRs.

**Pull request closed without merging**